sql/core/src/test/scala/org/apache/spark/sql/MapAggregateQuerySuite.scala

```diff
@@ -0,0 +1,124 @@
+  import testImplicits._
+
+  private val table = "map_aggregate_test"
+  private val col1 = "col1"
+  private val col2 = "col2"
+  private def query(numBins: Int): DataFrame = {
+    sql(s"SELECT map_aggregate($col1, $numBins), map_aggregate($col2, $numBins) FROM $table")
+  }
+
+  test("null handling") {
+    withTempView(table) {
+      val schema = StructType(Seq(StructField(col1, StringType), StructField(col2, DoubleType)))
+      // Empty input row
+      val rdd1 = spark.sparkContext.parallelize(Seq(Row(null, null)))
+      spark.createDataFrame(rdd1, schema).createOrReplaceTempView(table)
+      checkAnswer(query(numBins = 2), Row(Map.empty, Map.empty))
+
+      // Add some non-empty row
+      val rdd2 = spark.sparkContext.parallelize(Seq(Row(null, 3.0D), Row("a", null)))
+      spark.createDataFrame(rdd2, schema).createOrReplaceTempView(table)
+      checkAnswer(query(numBins = 2), Row(Map(("a", 1)), Map((3.0D, 1))))
+    }
+  }
+
+  test("returns empty result when ndv exceeds numBins") {
+    withTempView(table) {
+      spark.sparkContext.makeRDD(
+        Seq(("a", 4), ("d", 2), ("c", 4), ("b", 1), ("a", 3), ("a", 2)), 2).toDF(col1, col2)
+        .createOrReplaceTempView(table)
+      checkAnswer(query(numBins = 4), Row(
+        Map(("a", 3), ("b", 1), ("c", 1), ("d", 1)),
+        Map((1.0D, 1), (2.0D, 2), (3.0D, 1), (4.0D, 2))))
+      // One partial exceeds numBins during update()
+      checkAnswer(query(numBins = 2), Row(Map.empty, Map.empty))
+      // Exceeding numBins during merge()
+      checkAnswer(query(numBins = 3), Row(Map.empty, Map.empty))
+    }
+  }
+
+  test("multiple columns of different types") {
+    def queryMultiColumns(numBins: Int): DataFrame = {
+      sql(
+        s"""
+           |SELECT
+           |  map_aggregate(c1, $numBins),
+           |  map_aggregate(c2, $numBins),
+           |  map_aggregate(c3, $numBins),
+           |  map_aggregate(c4, $numBins),
+           |  map_aggregate(c5, $numBins),
+           |  map_aggregate(c6, $numBins),
+           |  map_aggregate(c7, $numBins),
+           |  map_aggregate(c8, $numBins),
+           |  map_aggregate(c9, $numBins),
+           |  map_aggregate(c10, $numBins)
+           |FROM $table
+        """.stripMargin)
+    }
+
+    val ints = Seq(5, 3, 1)
+    val doubles = Seq(1.0D, 3.0D, 5.0D)
+    val dates = Seq("1970-01-01", "1970-02-02", "1970-03-03")
+    val timestamps = Seq("1970-01-01 00:00:00", "1970-01-01 00:00:05", "1970-01-01 00:00:10")
+    val strings = Seq("a", "bb", "ccc")
+
+    val data = ints.indices.map { i =>
+      (ints(i).toByte,
+        ints(i).toShort,
+        ints(i),
+        ints(i).toLong,
+        doubles(i).toFloat,
+        doubles(i),
+        Decimal(doubles(i)),
+        Date.valueOf(dates(i)),
+        Timestamp.valueOf(timestamps(i)),
+        strings(i))
+    }
+    withTempView(table) {
+      data.toDF("c1", "c2", "c3", "c4", "c5", "c6", "c7", "c8", "c9", "c10")
```

Comment:
Can you use `def createDataFrame(rowRDD: RDD[Row], schema: StructType)`, instead of using implicit schema inference? 

We need an explicit schema definition when we verifying different data types.